File: DCP-1825/solution.swift
// Problem 1825
// Circular Queue

import Foundation

class Graph {
    private let vertices: Int
    private var adj: [[Int]]
    private var inDegree: [Int]
    
    init(_ v: Int) {
        vertices = v
        adj = Array(repeating: [], count: v)
        inDegree = Array(repeating: 0, count: v)
    }
    
    func addEdge(_ v: Int, _ w: Int) {
        adj[v].append(w)
        inDegree[w] += 1
    }
    
    private func isStronglyConnected() -> Bool {
        var visited = Array(repeating: false, count: vertices)
        
        // Find a vertex with non-zero degree
        var startVertex = 0
        for i in 0..<vertices {
            if !adj[i].isEmpty {
                startVertex = i
                break
            }
        }
        
        // DFS starting from startVertex
        dfs(startVertex, &visited)
        
        // Check if all non-zero degree vertices are visited
        for i in 0..<vertices {
            if !adj[i].isEmpty && !visited[i] {
                return false
            }
        }
        
        return true
    }
    
    private func dfs(_ v: Int, _ visited: inout [Bool]) {
        visited[v] = true
        for u in adj[v] {
            if !visited[u] {
                dfs(u, &visited)
            }
        }
    }
    
    func hasEulerianCycle() -> Bool {
        if !isStronglyConnected() {
            return false
        }
        
        // Check if in degree equals out degree for every vertex
        for i in 0..<vertices {
            if adj[i].count != inDegree[i] {
                return false
            }
        }
        
        return true
    }
}

func canChainWords(_ words: [String]) -> Bool {
    let graph = Graph(26) // 26 letters
    
    // Create edges between first and last characters
    for word in words {
        let first = Int(word.first!.asciiValue! - Character("a").asciiValue!)
        let last = Int(word.last!.asciiValue! - Character("a").asciiValue!)
        graph.addEdge(first, last)
    }
    
    return graph.hasEulerianCycle()
}

File: DCP-1825/solution.kt
// Problem 1825

class Graph(private val vertices: Int) {
    private val adj = Array(vertices) { mutableListOf<Int>() }
    private val inDegree = IntArray(vertices)
    
    fun addEdge(v: Int, w: Int) {
        adj[v].add(w)
        inDegree[w]++
    }
    
    private fun dfs(v: Int, visited: BooleanArray) {
        visited[v] = true
        for (u in adj[v]) {
            if (!visited[u]) {
                dfs(u, visited)
            }
        }
    }
    
    private fun isStronglyConnected(): Boolean {
        val visited = BooleanArray(vertices)
        
        // Find a vertex with non-zero degree
        val startVertex = (0 until vertices).find { adj[it].isNotEmpty() } ?: return true
        
        // DFS starting from startVertex
        dfs(startVertex, visited)
        
        // Check if all non-zero degree vertices are visited
        return (0 until vertices).none { adj[it].isNotEmpty() && !visited[it] }
    }
    
    fun hasEulerianCycle(): Boolean {
        if (!isStronglyConnected()) return false
        
        // Check if in degree equals out degree for every vertex
        return (0 until vertices).all { adj[it].size == inDegree[it] }
    }
}

fun canChainWords(words: List<String>): Boolean {
    val graph = Graph(26) // 26 letters
    
    // Create edges between first and last characters
    words.forEach { word ->
        val first = word.first() - 'a'
        val last = word.last() - 'a'
        graph.addEdge(first, last)
    }
    
    return graph.hasEulerianCycle()
}